lib/lower/ASTBuilder.cpp

```diff
@@ -0,0 +1,783 @@
+    evallist.pop_back();
+  }
+
+  AST::Program *pgm;
+  std::list<AST::FunctionLikeUnit> *funclist{nullptr};
+  std::vector<std::list<AST::Evaluation> *> evallist;
+  std::vector<AST::ParentType> parents;
+};
+
+template <typename A>
+constexpr bool hasErrLabel(const A &stmt) {
+  if constexpr (std::is_same_v<A, Pa::ReadStmt> ||
+                std::is_same_v<A, Pa::WriteStmt>) {
+    for (const auto &control : stmt.controls) {
+      if (std::holds_alternative<Pa::ErrLabel>(control.u)) {
+        return true;
+      }
+    }
+  }
+  if constexpr (std::is_same_v<A, Pa::WaitStmt> ||
+                std::is_same_v<A, Pa::OpenStmt> ||
+                std::is_same_v<A, Pa::CloseStmt> ||
+                std::is_same_v<A, Pa::BackspaceStmt> ||
+                std::is_same_v<A, Pa::EndfileStmt> ||
+                std::is_same_v<A, Pa::RewindStmt> ||
+                std::is_same_v<A, Pa::FlushStmt>) {
+    for (const auto &spec : stmt.v) {
+      if (std::holds_alternative<Pa::ErrLabel>(spec.u)) {
+        return true;
+      }
+    }
+  }
+  if constexpr (std::is_same_v<A, Pa::InquireStmt>) {
+    for (const auto &spec : std::get<std::list<Pa::InquireSpec>>(stmt.u)) {
+      if (std::holds_alternative<Pa::ErrLabel>(spec.u)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+template <typename A>
+constexpr bool hasEorLabel(const A &stmt) {
+  if constexpr (std::is_same_v<A, Pa::ReadStmt> ||
+                std::is_same_v<A, Pa::WriteStmt>) {
+    for (const auto &control : stmt.controls) {
+      if (std::holds_alternative<Pa::EorLabel>(control.u)) {
+        return true;
+      }
+    }
+  }
+  if constexpr (std::is_same_v<A, Pa::WaitStmt>) {
+    for (const auto &waitSpec : stmt.v) {
+      if (std::holds_alternative<Pa::EorLabel>(waitSpec.u)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+template <typename A>
+constexpr bool hasEndLabel(const A &stmt) {
+  if constexpr (std::is_same_v<A, Pa::ReadStmt> ||
+                std::is_same_v<A, Pa::WriteStmt>) {
+    for (const auto &control : stmt.controls) {
+      if (std::holds_alternative<Pa::EndLabel>(control.u)) {
+        return true;
+      }
+    }
+  }
+  if constexpr (std::is_same_v<A, Pa::WaitStmt>) {
+    for (const auto &waitSpec : stmt.v) {
+      if (std::holds_alternative<Pa::EndLabel>(waitSpec.u)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool hasAltReturns(const Pa::CallStmt &callStmt) {
+  const auto &args{std::get<std::list<Pa::ActualArgSpec>>(callStmt.v.t)};
+  for (const auto &arg : args) {
+    const auto &actual{std::get<Pa::ActualArg>(arg.t)};
+    if (std::holds_alternative<Pa::AltReturnSpec>(actual.u)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Determine if `callStmt` has alternate returns and if so set `e` to be the
+/// origin of a switch-like control flow
+void altRet(AST::Evaluation &e, const Pa::CallStmt *callStmt,
+            AST::Evaluation *cstr) {
```

Comment:
Here and in following functions, if the `cstr` argument pointer can't be null, please use a reference instead.

Reply:
`cstr` can be null see: https://github.com/flang-compiler/f18/blob/16ab9dacd8993694ace6c1c85dd8a4763ba89c21/lib/lower/ASTBuilder.cpp#L649
The makes a differences here:
https://github.com/flang-compiler/f18/blob/16ab9dacd8993694ace6c1c85dd8a4763ba89c21/include/flang/lower/ASTBuilder.h#L224

This is because `cstr` is the current construct and at the top level of a procedure/program there is no such "current construct".

I have added comments saying that.